pkg/pub_package_reader/lib/src/pubspec_content_override.dart

```diff
@@ -0,0 +1,228 @@
+  'meta_types_json_generator',
+  'meta_types_redux',
+  'meta_types_redux_generator',
+  'mocktail',
+  'mocktail_image_network',
+  'polymer_interop',
+};
+
+/// The strict version parsing fix will be part of this SDK release.
+final _maxSdkVersion = Version(2, 17, 0);
+final _defaultSdkVersionRange = VersionRange(
+  min: Version(2, 12, 0),
+  includeMin: true,
+  max: _maxSdkVersion,
+  includeMax: false,
+);
+
+/// Override pubspec.yaml if needed:
+///
+/// - If the archive was created before 2022-01-01, we may need to update the
+///   version number, version constraint and SDK constraints, as `pub_semver`
+///   accepted separator characters other than `.`.
+///   https://github.com/dart-lang/pub_semver/pull/63
+String overridePubspecYamlIfNeeded({
+  required String pubspecYaml,
+  required DateTime published,
+}) {
+  // quick checks to skip overrides
+  if (published.year >= 2022) {
+    return pubspecYaml;
+  }
+
+  try {
+    return _fixupBrokenVersionAndConstraints(pubspecYaml);
+  } on FormatException catch (e, st) {
+    _logger.info('Unable to parse pubspec.', e, st);
+  }
+  // fallback: don't override
+  return pubspecYaml;
+}
+
+/// Fixes broken version and version constraints in [pubspecYaml] and returns a new YAML string.
+///
+/// Prior to [pub_semver#63] version numbers and version constraints were accepted with any
+/// character as separator, instead of requiring  `.` (dot). This occurred because the regular expression
+/// in `package:pub_semver` didn't escape dots as `\.`, but just contained `.` (dot, meaning any character). (see https://github.com/dart-lang/pub_semver/commit/cee044a3dc8)
+///
+/// Thus, package versions published prior to 2022-01-01, may contain _version numbers_,
+/// _version constraints_ and _SDK constraints_ that don't use `.` (dot) as separator. As removing
+/// these _package versions_ would break existing users, this function will rewrite the version numbers
+/// and version constraints to have a valid format equivalent to how older versions of `pub_semver`
+/// would have interpreted the version number / version constraints.
+///
+/// [pub_semver#63]: https://github.com/dart-lang/pub_semver/pull/63
+String _fixupBrokenVersionAndConstraints(String pubspecYaml) {
+  final root = yaml.loadYaml(pubspecYaml);
+  if (root is! Map) {
+    _logger.warning(
+        'Unable to parse YAML for package, in _fixupBrokenVersionAndConstraints!');
+    return pubspecYaml; // return original as no changes can be made
+  }
+
+  // sanity check on the name again, now with parsed yaml
+  final name = root['name'];
+  if (name is! String || !_packagesWithBadVersions.contains(name)) {
+    return pubspecYaml; // return original as no changes needed
+  }
+
+  final editor = YamlEditor(pubspecYaml);
+  var hasBeenUpdated = false;
+  for (final c in _detectVersionEditCandidates(root)) {
+    final updated = c.value
+        // Version constraints may have more than one version expression, separated by space.
```

Comment:
Is whitespace separation required? That is what this comment says, and the code relies on it.

Reply:
Since the applied regexp is between `^` and `$`, I think we need to split by whitespaces. Or am I missing something here that makes this not necessary?

Reply:
But you assume there is whitespace, which there might not be in the general case... right?

Reply:
Why don't we copy/paste the functions from `pub_semver` then we're sure we have the same logic.
Here they are, ported to not depend on anything from `pub_semver` other than they create instances of `Version` and `VersionConstraint`.

```dart
import 'package:pub_semver/pub_semver.dart';

/// Parses a comparison operator ("<", ">", "<=", or ">=") at the beginning of
/// a string.
final _startComparison = RegExp(r'^[<>]=?');

/// The "compatible with" operator.
const _compatibleWithChar = '^';

/// Regex that matches a version number at the beginning of a string.
final _startVersion = RegExp(r'^' // Start at beginning.
    r'(\d+)\.(\d+)\.(\d+)' // Version number.
    r'(-([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Pre-release.
    r'(\+([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?'); // Build.

/// Like [_startVersion] but matches the entire string.
final _completeVersion = RegExp('${_startVersion.pattern}\$');

/// Creates a new [Version] by parsing [text].
Version _parsePossiblyBrokenVersion(String text) {
  final match = _completeVersion.firstMatch(text);
  if (match == null) {
    throw FormatException('Could not parse "$text".');
  }

  try {
    var major = int.parse(match[1]!);
    var minor = int.parse(match[2]!);
    var patch = int.parse(match[3]!);

    var preRelease = match[5];
    var build = match[8];

    return Version(major, minor, patch, pre: preRelease, build: build);
  } on FormatException {
    throw FormatException('Could not parse "$text".');
  }
}

/// Parses a version constraint.
///
/// Following rules from `package:pub_semver <= 2.1.0`.
VersionConstraint _parsePossiblyBrokenVersionConstraint(String text) {
  var originalText = text;

  void skipWhitespace() {
    text = text.trim();
  }

  skipWhitespace();

  // Handle the "any" constraint.
  if (text == 'any') return VersionConstraint.any;

  // Try to parse and consume a version number.
  Version? matchVersion() {
    var version = _startVersion.firstMatch(text);
    if (version == null) return null;

    text = text.substring(version.end);
    return Version.parse(version[0]!);
  }

  // Try to parse and consume a comparison operator followed by a version.
  VersionRange? matchComparison() {
    var comparison = _startComparison.firstMatch(text);
    if (comparison == null) return null;

    var op = comparison[0];
    text = text.substring(comparison.end);
    skipWhitespace();

    var version = matchVersion();
    if (version == null) {
      throw FormatException('Expected version number after "$op" in '
          '"$originalText", got "$text".');
    }

    switch (op) {
      case '<=':
        return VersionRange(max: version, includeMax: true);
      case '<':
        return VersionRange(
            max: version, includeMax: false, alwaysIncludeMaxPreRelease: true);
      case '>=':
        return VersionRange(min: version, includeMin: true);
      case '>':
        return VersionRange(min: version, includeMin: false);
    }
    throw FallThroughError();
  }

  // Try to parse the "^" operator followed by a version.
  VersionConstraint? matchCompatibleWith() {
    if (!text.startsWith(_compatibleWithChar)) return null;

    text = text.substring(_compatibleWithChar.length);
    skipWhitespace();

    var version = matchVersion();
    if (version == null) {
      throw FormatException('Expected version number after '
          '"$_compatibleWithChar" in "$originalText", got "$text".');
    }

    if (text.isNotEmpty) {
      throw FormatException('Cannot include other constraints with '
          '"$_compatibleWithChar" constraint in "$originalText".');
    }

    return VersionConstraint.compatibleWith(version);
  }

  var compatibleWith = matchCompatibleWith();
  if (compatibleWith != null) return compatibleWith;

  Version? min;
  var includeMin = false;
  Version? max;
  var includeMax = false;

  for (;;) {
    skipWhitespace();

    if (text.isEmpty) break;

    var newRange = matchVersion() ?? matchComparison();
    if (newRange == null) {
      throw FormatException('Could not parse version "$originalText". '
          'Unknown text at "$text".');
    }

    if (newRange.min != null) {
      if (min == null || newRange.min! > min) {
        min = newRange.min;
        includeMin = newRange.includeMin;
      } else if (newRange.min == min && !newRange.includeMin) {
        includeMin = false;
      }
    }

    if (newRange.max != null) {
      if (max == null || newRange.max! < max) {
        max = newRange.max;
        includeMax = newRange.includeMax;
      } else if (newRange.max == max && !newRange.includeMax) {
        includeMax = false;
      }
    }
  }

  if (min == null && max == null) {
    throw const FormatException('Cannot parse an empty string.');
  }

  if (min != null && max != null) {
    if (min > max) return VersionConstraint.empty;
    if (min == max) {
      if (includeMin && includeMax) return min;
      return VersionConstraint.empty;
    }
  }

  return VersionRange(
      min: min, includeMin: includeMin, max: max, includeMax: includeMax);
}
```

This is an adoption of what I used when detecting if we had broken versions in the first place.

If a version doesn't satisfy the new pattern (which isn't in a published version of pub_semver yet), then we need to parse with these functions. And use the canonicalized constraint, which `toString()` should create, because I don't pass `text` to the constructor.

Reply:
Updated. I think the parsing with the spaces was equivalent, but you are right, better to use the snapshot of the pub_semver library. I've put it into a separate file, just in case we have more of these in the future, let's not mix them.

Reply:
> Afaik spaces are not required

TIL. yeah, better to stick with the existing code.